### PR TITLE
Fixed #29899 -- Made autodetector use model states instead of model classes.

### PR DESCRIPTION
**Pull request by David-Wobrock**
Ticket https://code.djangoproject.com/ticket/29899

Hey :wave: ,
I'd suggest reviewing what I did commit-by-commit, in order to be less overwhelmed. The commits probably will need to be squashed before merging, to have one or more meaningful and atomic commits.

The goal is pretty clearly stated through the ticket description.
I started the work off @charettes' https://github.com/django/django/compare/master...charettes:autodetector-model-states, tried to get my head around and work myself through to get it to work.

The things I think need the most review attention and which might be brittle are:
* handling swappables
* filling the field names of the relations (the logic of choosing the primary key or a `{model}_ptr`)
* basically all the methods added to `ProjectState` (see the commit named _Add and call method of ProjectState to fill relations_) to which I'm rather unsure

I basically iterated to what seemed to make sense by making the unit tests pass one-by-one. However, I cannot guarantee that there isn't a regression on an untested corner case :confused:

### CODE REVIEW

File: django/db/models/fields/related.py [hunk 1]
```diff
@@ -34,6 +34,19 @@
 RECURSIVE_RELATIONSHIP_CONSTANT = 'self'
 
 
+def resolve_model_key(app_label, model_name, relation):
```
Comment (charettes) [inline django/db/models/fields/related.py:37]: I think you could use `django.db.models.migrations.operation.resolve_relation` for this purpose. Could be worth moving to `django.db.models.migrations.utils`.

File: django/db/migrations/autodetector.py [hunk 2]
```diff
@@ -225,14 +226,11 @@ def _generate_through_model_map(self):
...
-                    through_key = (
-                        old_field.remote_field.through._meta.app_label,
-                        old_field.remote_field.through._meta.model_name,
-                    )
-                    self.through_users[through_key] = (app_label, old_model_name, field_name)
+                if hasattr(field, 'remote_field') and getattr(field.remote_field, 'through', None):
+                    through_key = resolve_model_key(app_label, model_name, field.remote_field.through)
+                    through_model_state = self.from_state.models[through_key]
+                    if not through_model_state.options.get("auto_created"):
+                        self.through_users[through_key] = (app_label, old_model_name, field_name)
```
Comment (charettes) [inline django/db/migrations/autodetector.py:234]: Are you sure this branch is ever skipped? AFAIK `auto_created` models are not part `ProjectState.models` entries.

File: django/db/migrations/autodetector.py [hunk 3]
```diff
@@ -539,9 +537,15 @@ def generate_created_models(self):
...
-                        related_fields[field.name] = field
+                    if getattr(field.remote_field, "through", None):
+                        through_app_label, through_model_name = resolve_model_key(
+                            app_label,
+                            model_name,
+                            field.remote_field.through,
+                        )
+                        through_model_state = self.to_state.models[through_app_label, through_model_name]
+                        if not through_model_state.options.get("auto_created"):
+                            related_fields[field_name] = field
```
Comment (charettes) [inline django/db/migrations/autodetector.py:553]: Ditto, I think this will always be `True`.

File: django/db/migrations/autodetector.py [hunk 4]
```diff
@@ -1066,23 +1072,21 @@ def generate_removed_constraints(self):
...
-                field.remote_field.through._meta.app_label,
-                field.remote_field.through._meta.object_name,
-                None,
-                True,
-            ))
+        if getattr(field.remote_field, "through", None):
+            through_app_label, through_object_name = resolve_model_key(app_label, model_name, field.remote_field.model)
+            through_model_state = project_state.models[through_app_label, through_object_name]
+            if not through_model_state.options.get("auto_created"):
+                dependencies.append((through_app_label, through_object_name, None, True))
```
Comment (charettes) [inline django/db/migrations/autodetector.py:1090]: Ditto, I'm pretty sure this will always be `True`.

File: django/db/migrations/state.py [hunk 5]
```diff
@@ -188,6 +193,89 @@ def _reload(self, related_models):
...
+            self.relations[model_key] = self.relations[concretes[model_key]]
+
+    def fill_remote_field_name(self, field):
+        if self.is_swappable_model(field.remote_field.model):
+            field.remote_field.field_name = "id"
+        else:
+            try:
+                model_state = self.get_concrete_model_state(field.remote_field.model)
+                base = model_state.bases[0]
+                if not isinstance(base, str) and base.__name__ == "Model":
```
Comment (charettes) [inline django/db/migrations/state.py:234]: This seem a bit arbitrary, is it the original logic? I'd type check against `base` instead.

File: django/db/migrations/state.py [hunk 6]
```diff
@@ -188,6 +193,89 @@ def _reload(self, related_models):
...
+                    field.remote_field.field_name = model_state.get_primary_key_field_name()
+                    model_state.all_fields.add(field.remote_field)
+                else:
+                    base_model_state = self.get_concrete_model_state(base)
+                    field.remote_field.field_name = '%s_ptr' % base_model_state.name_lower
+                    base_model_state.all_fields.add(field.remote_field)
+            except KeyError:
+                pass  # not migrated
```
Comment (charettes) [inline django/db/migrations/state.py:242]: I'd be great to have bit more clarity about which failure case we're handling here with `KeyError`. This seems like a complex routine wrapped in a catch all.

File: django/db/migrations/state.py [hunk 7]
```diff
@@ -360,6 +448,7 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
         self.app_label = app_label
         self.name = name
         self.fields = fields
+        self.all_fields = {f for _, f in fields}
```
Comment (charettes) [inline django/db/migrations/state.py:451]: So this is meant to store inherited fields as well?

File: django/db/migrations/state.py [hunk 8]
```diff
@@ -188,6 +193,89 @@ def _reload(self, related_models):
...
+                    through = getattr(remote_field, 'through', None)
+                    if through and not model_state.options.get('auto_created'):
+                        through_model_key = resolve_model_key(*model_key, through)
+                        if through_model_key[0] not in real_apps:
+                            through_model_key = concretes[through_model_key]
+                        self.relations[through_model_key][model_key].append((field_name, field))
+        for model_key in proxies:
+            self.relations[model_key] = self.relations[concretes[model_key]]
+
+    def fill_remote_field_name(self, field):
```
Comment (charettes) [inline django/db/migrations/state.py:227]: I wonder if this is something we could get rid of, `RenameField.state_forwards` has to do a bit of gymnastic to handle that as well. Do you remember what breaks when `field_name` is not assigned properly?

File: django/db/migrations/state.py [hunk 9]
```diff
@@ -87,6 +90,8 @@ def __init__(self, models=None, real_apps=None):
         # Apps to include from main registry, usually unmigrated ones
         self.real_apps = real_apps or []
         self.is_delayed = False
+        # Attributes filled after calling resolve_fields_and_relations
+        self.relations = None
```
Comment (charettes) [inline django/db/migrations/state.py:94]: So that's kind of a reverse relation map similar to `models.Options.related_objects`?

File: django/db/migrations/state.py (hunk 8 above)
Comment (David-Wobrock) [reply django/db/migrations/state.py:227]: Not much would fail on the autodetector logic actually. It is more that multiple tests are currently asserting that the remote field name. See https://github.com/django/django/pull/12688/files#diff-c11e6432df7086eda3dfb9ab8e5b2839R1691 or https://github.com/django/django/pull/12688/files#diff-c11e6432df7086eda3dfb9ab8e5b2839R1790 among others (6 tests are failing).

If you think that these tests can be removed or substituted, that would definitely simplify the logic and remove code. However, I wouldn't know how to meaningfully replace the `field_name` asserts in the `test_unmanaged_custom_pk` test for instance (https://github.com/django/django/blob/master/tests/migrations/test_autodetector.py#L1735-L1747)

File: django/db/migrations/state.py (hunk 7 above)
Comment (David-Wobrock) [reply django/db/migrations/state.py:451]: It is meant to store "target"/remote fields. Meaningly, `ManyToOneRel` instances.
It is actually only used in one place => https://github.com/django/django/pull/12688/files#diff-a71b173d008afea1acd5318ef509bb7cR486
Else, the test https://github.com/django/django/blob/master/tests/migrations/test_autodetector.py#L1116-L1145 fails because we do not detect the dependency between both models, and therefore the dependency between both apps.
See the error
```
======================================================================
FAIL: test_rename_model_reverse_relation_dependencies (migrations.test_autodetector.AutodetectorTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "django/tests/migrations/test_autodetector.py", line 1148, in test_rename_model_reverse_relation_dependencies
    self.assertMigrationDependencies(changes, 'testapp', 0, [('otherapp', '__first__')])
  File "django/tests/migrations/test_autodetector.py", line 502, in assertMigrationDependencies
    self.repr_changes(changes, include_dependencies=True),
AssertionError: Migration dependencies mismatch for testapp.auto_1 (expected [('otherapp', '__first__')]):
  testapp:
    auto_1
      <RenameModel  old_name='EntityA', new_name='RenamedEntityA'>
      Dependencies:
        None
```

Two things:
1. I guess that this is not the right approach. I should be able to somehow leverage the `relations` attribute.
2. I don't know if this logic (of checking relations when iterating through fields) should also be applied in other methods. Like model creation detection. But probably yes, so that is missing from my PR currently

File: django/db/migrations/state.py (hunk 6 above)
Comment (David-Wobrock) [reply django/db/migrations/state.py:242]: The 3 tests that fail because we are not able to resolve the concrete model are:
* `test_first_dependency`
* `test_last_dependency`
* `test_swappable_changed`

Basically because the FK fields point a model that does not exist in the state `"migrations.UnmigratedModel"` (`https://github.com/django/django/blob/master/tests/migrations/test_autodetector.py#L352`) or the model is inhering from an unknown model in the state `AbstractBaseUse` (https://github.com/django/django/blob/master/tests/migrations/test_autodetector.py#L441-L444)

These can probably be handled more elegantly

File: django/db/migrations/state.py (hunk 5 above)
Comment (David-Wobrock) [reply django/db/migrations/state.py:234]: This comes from the default bases we attribute on ModelState: https://github.com/django/django/blob/master/django/db/migrations/state.py#L366 
But you are right that it seems that it isn't needed actually.

`if not isinstance(base, str):` seems to be enough. I think I had troubles making tests pass when I didn't add catching the KeyError.

File: django/db/migrations/autodetector.py (hunk 2 above)
Comment (David-Wobrock) [reply django/db/migrations/autodetector.py:234]: You're most probably right :+1: I'll remove those conditions

File: django/db/migrations/state.py (hunk 9 above)
Comment (David-Wobrock) [reply django/db/migrations/state.py:94]: Somewhat, yes, but not exactly.
To illustrate, with models like
`app.User` and `app.Author with owner = ForeignKey(app.User)`

`relations` will contain something like
```
{
  ("app", "user"): {
    ("app", "author"): [
      ("owner", ForeignKey)
    ]
  }
}
```
whereas `models.Options.related_objects` contains, for a model, the reverse relation `ManyToOneRel` (and others) objects

**Comment by David-Wobrock** [comment]
> Let me start by thanking you for reviving this branch and taking the time to adapt it. I left a few comments and questions but over all that looks promising.
> 
> I think the commits could be reorganized bit but I think the bulk of the work is there. I like your `ProjectState.relation` approach. One thing that will need to be considered though is that it should likely be invalidated when `reload_model` is called even if the latter is not used used the auto-detector. That when both a forward and reverse relation map might come handy.
> 
> You might want to peak at recent changes `django.db.models.migrations.operations.utils` to see if anything there might be useful to move to `django.db.models.migrations.utils` for reuse.

Thanks a lot for the review! I took the easy changes into account in the last 4 commits (I had to rebase and force-push, sorry :/)
I answered the other comments and hope that you will be able to point me to the right direction.

Thanks again :)

**Comment by David-Wobrock** [comment]
Hi @charettes 
I rebased my branch against the latest `master` and squashed some commits.
I also tried to simplify and remove some logic here and there.
Most notably, I removed the logic filling the FK fields remote name, which was duplicating logic that was inside the rendered FK. However, this meant changing the assertions of some tests. I don't know how critical it is to test it in the autodetector itself, but it seemed not necessarily relevant there.

Looking forward to more reviews :)

**Review by charettes — COMMENTED** [review]
Hello David, sorry for how long it took to provide feedback to your PR and thank you for your contribution in the mean time.

This is really excellent work in my opinion and I appreciate the per-commit beakdown as it made the review more approachable.

One thing that concerns me though is using `Field.model` as a mean to store a back reference to owner of the field. I fear it might cause issues down the line where we could forget to rename yet another attribute when altering state.

Would it be possible to adapt the relations data structure to store `(model: str, field: Field)` instead to avoid altering `Field` and ensuring it's kept in sync down the line?

File: django/db/migrations/state.py [hunk 10]
```diff
@@ -367,7 +429,7 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
         self.managers = managers or []
         for name, field in self.fields.items():
             # Sanity-check that fields are NOT already bound to a model.
-            if hasattr(field, 'model'):
+            if hasattr(field, 'model') and not isinstance(field.model, str):
```
[the remark this answers is not in the document]
Comment (David-Wobrock) [reply django/db/migrations/state.py]: Thanks for the review. I added a small commit.
I'd be grateful if you could review that one :)

However I think I'm missing something here.
The fact that I had to modify the tests that much (in the PR generally) bothers me - they are not asserting what they should anymore. However I'm not sure how to achieve having access to the related field name.
One typical example would be https://github.com/django/django/pull/12688/files#diff-506caa00017053ff8278de6efc2e59cc0c5cea22da9461482bdf16a9fc50af9eL1763-R1791 where we try to assert the target field. But this test now feels like it lost its value.

[the remark this answers is not in the document]
Comment (charettes) [reply django/db/migrations/state.py]: My gut feeling is that it's normal that these test assertions lost a bit of their original _intent_ now we're not using rendered models anymore.

The reason for that is that `.field_name` gets assigned on model rendering which we completely avoid doing here. The fact _rendered_ model fields were making their way into operations breaks the `ModelState.__init__` expectations and was really just an implementation detail.

Looking more at the #23415 and the resulting patch 215aa4f53b6bbd07d5c1eecfa94e7fcd00da813e these particular assertions had little to do with the issue of unmanaged models not including their fields in the first place.

TL;DR I think this is fine.

[the remark this answers is not in the document]
Comment (David-Wobrock) [reply django/db/migrations/state.py]: Cool, thanks a lot for the input :+1:

**Review by charettes — APPROVED** [review]
After another review I think this is ready to go!

It'd be great to run this branch against large generated project with a few cross referencing models to see how it behaves. After all we were hoping this might help a bit there.

I can't find anything of this kind in https://github.com/django/djangobench/tree/master/djangobench/benchmarks but IIRC @MarkusH had such a script available.

**Comment by MarkusH** [comment]
> After another review I think this is ready to go!
> 
> It'd be great to run this branch against large generated project with a few cross referencing models to see how it behaves. After all we were hoping this might help a bit there.
> 
> I can't find anything of this kind in https://github.com/django/djangobench/tree/master/djangobench/benchmarks but IIRC @MarkusH had such a script available.

I can only agree with Simon. This is excellent work. Thank you.

As part of issue [#24743](https://code.djangoproject.com/ticket/24743) I created https://github.com/MarkusH/django-migrations-benchmark .

Unfortunately, I lack the time to have a close look at the changes.

I don't see a full discussion around Simon's comment in https://github.com/django/django/pull/12688#pullrequestreview-400404244 . Is there a chance to move the tracking of which field refers to which other field onto `StateApps` instead? That would allow for eventually moving the tracking of the reference outside of the migration framework to `django.apps.apps.Apps`. Which in turn would allow us to use `ModelStates` for applying migrations. But that's by no means a blocker at this point. If Simon says it's good to go, let's get the chance merged.

**Comment by David-Wobrock** [comment]
Hi @charettes, hi @MarkusH 

Thanks a lot for the reviews! :)

> I don't see a full discussion around Simon's comment in #12688 (review) 

In the PR, I first added a `model` attribute on the fly to the `Field` instances, but it was removed and instead, we iterate directly in the only place it's needed (if I recall correctly) here https://github.com/django/django/pull/12688/files#diff-50cc08301fdd233fd64196aa90bc684d96ef557d99af67692995559646de567aR1086-R1093
Did you mean that?

> Is there a chance to move the tracking of which field refers to which other field onto `StateApps` instead?

Sounds like a good idea! However I would prefer doing it in another ticket if possible. The hereby PR is open for quite some time, and I haven't looked at it for a few months. I'd need a bit of time to work on it again, but since I'm working on other projects that wouldn't be soon :)

> As part of issue #24743 I created https://github.com/MarkusH/django-migrations-benchmark .
Unfortunately, I lack the time to have a close look at the changes.

I took a quick look at the benchmark and reworked it so that `migrate ` passes on Django master (see https://github.com/MarkusH/django-migrations-benchmark/compare/master...David-Wobrock:rework-on-latest)
But I'm actually not sure how I would test the non-regression of this patch.
For the `migrate` speed, I don't see any significant change when trying on `master` and on this branch (and I didn't really expect it, since only the autodetector is affected by this patch, so makemigrations should only be affected)

Hope that helps and we'll be able to merge this soon 🚀

File: django/db/migrations/autodetector.py [hunk 11]
```diff
@@ -525,27 +533,25 @@ def generate_created_models(self):
...
-                    related_fields[field.name] = field
-                if getattr(field.remote_field, "through", None) and not field.remote_field.through._meta.auto_created:
-                    related_fields[field.name] = field
+                    if getattr(field.remote_field, "through", None):
+                        through_app_label, through_model_name = resolve_relation(
+                            field.remote_field.through,
+                            app_label,
+                            model_name,
+                        )
+                        related_fields[field_name] = field
```
Comment (charettes) [inline django/db/migrations/autodetector.py:554]: @David-Wobrock I checked out your branch to build an alternative to #13904 off model states and I noticed both `through_app_label` and `through_model_name` are unused which makes me believe something is off here.

Comment (felixxm) [reply django/db/migrations/autodetector.py:554]: Good catch :dart: 
```python
through_app_label, through_model_name = resolve_relation(...)
```
is unused here and in `generate_deleted_models()`. Do we need them? Also, related comments are outdated.

Comment (David-Wobrock) [reply django/db/migrations/autodetector.py]: Good catch indeed 👍 I don't know how this bit introduced itself in here, I guess missing attention. I removed it in the last commit.
Still, I haven't touched the PR for quite a while now, so I don't recall everything precisely - but it seems it's not necessary and shouldn't be there.

(I also rebased the PR on the latest `master` commit)

**Comment by David-Wobrock** [comment]
Hey again @charettes and @felixxm 👋 

As you pointed out, the bits of unused code was a little (harmless) mistake.

I rebased and ran again some manual tests of makemigrations to compare the generated migrations between the latest version on PyPI and this branch. I tried to cover as many cases as I could think of: creation, renaming, altering, deletion of concrete models, abstract models, proxy models, managed models, inheritance, FK fields, OneToOneFields with parent link, M2M, M2M with through and through_fields) and couldn't find any difference.

However, I'm still not 100% sure that I didn't miss any case, nor did I bench the performances of detecting a model change on a large project.

**Review by charettes — APPROVED** [review]
I ended up trying out the branch on two large projects with a few apps and models that are cross-referenced and there was a noticeable speed improvement when making some simple changes (e.g. reproducible 10% speed up when adding of removing a field) and similar speed ups when doing _complex_ changes (e.g. adding a few models that with cross-app references).

@felixxm at this point what is blocking us from moving forward with this PR? Do we want more detailed benchmarks against particular projects?

Beyond the apparent subtle (and likely heavily project specific) performance improvement the biggest gain is in term of having another part of migrations move away from unnecessary model rendering. Assuming this is merged and progress is made towards adapting schema editors to deal with model states instead of rendered model classes during GSoC we'd be able to remove all the complex and bug prone cache invalidation logic that was made necessary to optimize this code path.

The only location where model rendering would still be necessary would be in `RunPython` but that's relatively easy to do if the sole entrypoint is `apps.get_model` as you render a subset of the model graph discard it entirely once `database_forwards` is done with.

**Comment by felixxm** [comment]
> @felixxm at this point what is blocking us from moving forward with this PR? 

Nothing :+1: with your approval it is at the top of my list. Thanks :star:

File: django/db/migrations/state.py [hunk 12]
```diff
@@ -188,6 +192,63 @@ def _reload(self, related_models):
...
+
+        self.relations = defaultdict(lambda: defaultdict(list))
+        concretes, proxies = self._get_concrete_models_mapping_and_proxy_models()
+
+        real_apps = set(self.real_apps)
+        for model_key in concretes:
+            model_state = self.models[model_key]
+            for field_name, field in model_state.fields.items():
+                remote_field = field.remote_field
+                if remote_field:
```
Comment (cjerdonek) [inline django/db/migrations/state.py:208]: You can have a flatter function (less nesting) by doing:

```python
if not remote_field:
    continue
```

File: django/db/migrations/state.py [hunk 13]
```diff
@@ -188,6 +192,63 @@ def _reload(self, related_models):
...
+                    remote_model_key = resolve_relation(remote_field.model, *model_key)
+                    if remote_model_key[0] not in real_apps and remote_model_key in concretes:
+                        remote_model_key = concretes[remote_model_key]
+                    self.relations[remote_model_key][model_key].append((field_name, field))
+
+                    through = getattr(remote_field, 'through', None)
+                    if through:
```
Comment (cjerdonek) [inline django/db/migrations/state.py:215]: Ditto on flatter function.

File: django/db/migrations/utils.py [hunk 14]
```diff
@@ -15,3 +17,30 @@ def __eq__(self, other):
...
+
+
+def resolve_relation(model, app_label=None, model_name=None):
+    """
+    Turn a model class or model reference string and return a model tuple.
+
+    app_label and model_name are used to resolve the scope of recursive and
+    unscoped model relationship.
+    """
+    if isinstance(model, str):
```
Comment (cjerdonek) [inline django/db/migrations/utils.py:29]: Ditto by returning early.

**Review by cjerdonek — COMMENTED** [review]
A few comments to reduce nesting.

File: django/db/migrations/utils.py (hunk 14 above)
Comment (felixxm) [reply django/db/migrations/utils.py:29]: I would keep the current form.

**Comment by felixxm** [comment]
@David-Wobrock Many thanks for a great work :rocket: 
@charettes Thanks for all reviews and guidance :star: 

I pushed small edits and removed unused `ProjectState.concrete_apps`.

**Comment by felixxm** [comment]
buildbot, test on oracle.
